### ISSUE TITLE
Computing rbias on-the-fly for metadynamics

### ISSUE THREAD

**Issue by bodhivani** [openmm/openmm#4037]
Thanks for your response to my original question. I did look at getFreeEnergy, which makes so much more sense than reading the npy file from the biasDir as I was doing earlier.

I was hoping to save some time by not having to run a cubic spline every 1000 steps. I am assuming to compute forces, openMM is doing this computation internally-- is there a variable I can call during the simulation that does this?

I am also ok with approximating the integral just using the histogrammed bias values and using a more precise value for the bias function at the specific point in CV space at that time, so if you're just doing a local calculation for the bias that should be ok too.

---------

Original question:

tl;dr: I need the bias function every ~1000 steps to compute the on-the-fly reweighting factor. I also need the value of the bias at the current structure.

---------
I'm trying to compute on-the-fly the value of the reweighting factor as described in https://pubs.acs.org/doi/10.1021/jp504920s **

This is really pretty straightforward, it's some version of

constlog(int[over CV space] exp(V-gamma)/ int[over CV space] exp(const V))

(this is not exact, but you get the picture. Integral of some exponential function of the bias V over the CV space)

My current thought is to just grab the bias from the file saved in biasDir, take the appropriate functions, and add them up. However, this is a ~600x1000 file, so I'm assuming it's on gridpoints, but then computing the integral is not as direct as I'd thought. I could just set very finegrained gridpoints, but I assume this would make my actual simulation really slow (this already sounds larger than what I would set manually). Is there a variable I can call for the actual bias instantaneously to compute c(t) instead? For reweighting, I also need the instantaneous value of the bias at current structure.

**Instead of using eqns 13/14 as the paper suggests, I'm using eqn 3. This is how plumed does it as well, it's more statistically stable.

(sorry for the long-winded explanation!! I'm hoping it saves on more longwinded exchanges down the line)

**Comment by peastman** [comment]
The splines get refit every time the bias function is updated, meaning at whatever frequency you specify in the Metadynamics constructor.  The coefficients aren't available through any public API.

600 by 1000 seems like a really big table.  Is such high resolution really needed?  It's determined by the parameters you set for the BiasVariables.

**Comment by bodhivani** [comment]
Hmm ok, thanks! I will just use an approximation*-- simpson's rule on the grid should do it.

Regarding the large grid: I'm doing some pretty exploratory sampling on a CV without too much clear physical intuition (it's the latent space of a VAE), so I'm not sure exactly how far it's going to go. To be safe, I'm setting really large gridmin/gridmax. I'm not setting my own #bins, but I recall reading in the source code that it's set to 4*(grid range)/(gaussian width), which seems reasonable to me. Currently I get 130 ns/day (compared to 200 ns/day unbiased) so I'm not unhappy with it, and slightly wary of a coarser grid resulting in bad approximations for the bias. I'm happy to run an instance with a coarser grid to benchmark the speedup though.
